### ISSUE TITLE
General Travis macOS update ticket

### ISSUE THREAD

**Issue by dyfer** [supercollider/supercollider#5177]
I'd like to create a separate ticket, since I've recently worked on a number of issues that are interrelated. Specifics regarding each issue should be discussed in the separate ticket/PR, however I think we should use this thread so 1) we are aware of all the interrelations between the issues and 2) discuss the path forward, particularly WRT updating macOS Travis builds

# List of issues
- Current macOS Travis build fail (#5167) because sclang doesn't run when built with Qt 5.15.1 (#5175), although it runs fine with 5.15.0
  - ~~we don't have a fix for the Qt issue currently~~ EDIT: it seems to be a Qt bug
- We have [changed](https://github.com/supercollider/supercollider/issues/5033#issuecomment-696100369) Qt version in our builds between 3.11.0 (Qt 5.13.2) and 3.11.1 (Qt 5.15.0)
  - this caused SC to stop running on macOS 10.12; while we don't officially support it, it would be better not to make this change in a bugfix release
  - in relation to this I proposed that we start releasing a second, legacy build, that uses older Qt (#5174); IMO that would be a nice gift to users stuck on older systems 
- We have issues with CMake 3.18 (#5176), which is used by some newer Travis images (e.g. `xcode11.6`)
  - I have a [workaround](https://github.com/dyfer/supercollider/commit/8e3d71beec0881a945e9336aab24085b2a8bd246) for this, but I don't know what the root cause of the problem is
- Qt 5.13 and 5.14 cause problem with transparent windows when using custom color profile (#5033)

Not an issue directly, but a related point:
- We always pull the latest version of Qt for a given CI image on Travis
  - This caused the Qt change between bugfix releases
  - In the past we were more cautious about updating Qt
  - This also makes macOS builds taking longer to finish
  - I'd like to discuss the option of using newer Travis environments but not updating homebrew/Qt; `xcode11.6` image comes with Qt 5.15.0

# Before releasing 3.11.2 and 3.12
- We need to fix our macOS builds, either by fixing issue with Qt 5.15.1, or by using older Qt on Travis
- In 3.11.2 it would be nice to bring back support for older platforms that worked in 3.11.0

# Proposed scenario
Here's what I came up this, I'd love to get feedback on this
- in our 3.11 build, stop updating homebrew/qt 
  - this will revert support for older platform in 3.11.2
  - this will _not_ fix the issue with transparent windows in that release
  - there would be Qt "version regression" between 3.11.1 and 3.11.2; not sure if that's an issue
- update our build matrix in develop, for 3.12
  - stop updating homebrew/qt
  - switch main build to `xcode11.6`
    - this image has Qt 5.15.0 and CMake 3.18.0 installed
    - the issue with transparent windows would be solved
    - needs workaround or fix for CMake 3.18
    - no need to fix compatibility with Qt 5.15.1 before releasing SC 3.12
  - add legacy build, like in #5174
    - use `xcode10.1` image
    - use older version of Qt using a custom formula 
      - in my PR that's 5.9 which should make this run all the way back to macOS 10.10 (untested)
    - release the binary alongside our regular and regular+signed ones

I have a [test branch](https://github.com/supercollider/supercollider/compare/develop...dyfer:topic/travis-macos-speedup) with main build updated to `xcode11.6` image. I'm looking forward to some feedback on this, thanks!

**Comment by brianlheim** [comment]
> there would be Qt "version regression" between 3.11.1 and 3.11.2; not sure if that's an issue

yes, it was a major issue last time we did it. why not copy and use the 5.15.0 formula instead? that would be less drastic.

> no need to fix compatibility with Qt 5.15.1 before releasing SC 3.12

but anyone building locally on mac will hit this same issue; so if we don't fix compatibility we need to document it. is it possible to fix compatibility or can this only be fixed by Qt?

as i said in slack, not updating homebrew is a bad idea because we are just hiding from the problem and all other problems like it, not doing anything about it. people building themselves on mac are going to find the breakage too. we should be building with the latest packages as much as possible.

what i am seeing as most expedient is:
- try to resolve the issue with qt 5.15.1 ourselves if possible
- if not, then:
  - report it upstream to homebrew, they may consider making adding a 5.15.0 formula
  - document and announce this issue and the workaround
  - create a supercollider homebrew cask (not sure if this is the right terminology) with formulas for qt 5.9.x and 5.15.0
  - replace the oldest macOS build with a 5.9.x one in dev 
  - use the 5.15.0 formula on the other mac builds in dev and 3.11

we should probably also add the legacy build to 3.11 (if that's not too much work?)

**Comment by dyfer** [comment]
After private conversation with @brianlheim the next step will be updating our CI to use Qt 5.11.0 for the current builds and introduce the legacy build

**Comment by dyfer** [comment]
For completeness sake: we now have a workaround for Qt 5.11.1 (thanks @brianlheim !). After #5174 we should cherry pick it (and the additional platforms PR) into 3.11 so the CI passes and we are ready for another release. It will include both regular (macOS 10.13+) and legacy (macOS 10.10/10.12) builds, to remedy dropping support for macOS 10.12 in 3.11.1. 

IMO this ticket should be closed after that cherry pick.

**Comment by brianlheim** [comment]
that all makes sense to me! i think #5174 can be merged soon.

**Comment by dyfer** [comment]
In the end we have added legacy build only to the `3.11` branch (#5190). Other parts of this issue were addressed by #5174 and #5178. Outstanding issue with XCode 11 is tracked in #5188. I think this can be closed now.

**Comment by brianlheim** [comment]
thanks! this is the issue i was thinking of when i accidentally closed #5188
